mlflow/models/evaluation/default_evaluator.py

```diff
@@ -0,0 +1,651 @@
+
+    def _log_model_explainability(self):
+        if not self.evaluator_config.get("log_model_explainability", True):
+            return
+
+        if self.model_loader_module == "mlflow.spark":
+            # TODO: Shap explainer need to manipulate on each feature values,
+            #  but spark model input dataframe contains Vector type feature column
+            #  which shap explainer does not support.
+            #  To support this, we need expand the Vector type feature column into
+            #  multiple scaler feature columns and pass it to shap explainer.
+            _logger.warning(
+                "Logging model explainability insights is not currently supported for PySpark "
+                "models."
+            )
+            return
+
+        if self.model_type == "classifier" and not all(
+            [isinstance(label, (numbers.Number, np.bool_)) for label in self.label_list]
+        ):
+            _logger.warning(
+                "Skip logging model explainability insights because it requires all label "
+                "values to be Number type."
+            )
+            return
+
+        try:
+            import shap
+            import matplotlib.pyplot as pyplot
+        except ImportError:
+            _logger.warning(
+                "SHAP or matplotlib package is not installed, so model explainability insights "
+                "will not be logged."
+            )
+            return
+
+        if Version(shap.__version__) < Version("0.40"):
+            _logger.warning(
+                "Shap package version is lower than 0.40, Skip log model explainability."
+            )
+            return
+
+        is_multinomial_classifier = self.model_type == "classifier" and self.num_classes > 2
+
+        sample_rows = self.evaluator_config.get(
+            "explainability_nsamples", _DEFAULT_SAMPLE_ROWS_FOR_SHAP
+        )
+        algorithm = self.evaluator_config.get("explainability_algorithm", None)
+
+        truncated_feature_names = [truncate_str_from_middle(f, 20) for f in self.feature_names]
+        for i, truncated_name in enumerate(truncated_feature_names):
+            if truncated_name != self.feature_names[i]:
+                # For duplicated truncated name, attach "(f_{feature_index})" at the end
+                truncated_feature_names[i] = f"{truncated_name}(f_{i})"
+
+        truncated_feature_name_map = {
+            f: f2 for f, f2 in zip(self.feature_names, truncated_feature_names)
+        }
+
+        sampled_X = shap.sample(self.X, sample_rows)
+
+        if isinstance(sampled_X, pd.DataFrame):
+            # For some shap explainer, the plot will use the DataFrame column names instead of
+            # using feature_names argument value. So rename the dataframe column names.
+            sampled_X = sampled_X.rename(columns=truncated_feature_name_map, copy=False)
+
+        if algorithm:
+            supported_algos = ['exact', 'permutation', 'partition']
```

Comment:
Note:
in shap 0.40, the `shap.Explainer` argument `algo` , only when setting 'exact', 'permutation', 'partition', it works fine, the remaining algo (listed in shap doc) are all buggy in my test. So I limit them to be the 3 values.